File: svws-webclient/core/src/core/data/uv/UvRaum.ts
import { JavaLong } from '../../../java/lang/JavaLong';
import { JavaObject } from '../../../java/lang/JavaObject';
import { Class } from '../../../java/lang/Class';

export class UvRaum extends JavaObject {

	/**
	 * Die eindeutige ID des Raums (planungsspezifisch).
	 */
	public id: number = -1;

	/**
	 * Das Kürzel des Raums.
	 */
	public kuerzel: string = "";

	/**
	 * Das Datum, ab dem der Raum gültig ist.
	 */
	public gueltigVon: string = "";

	/**
	 * Das Datum, bis wann der Raum gültig ist. Ist kein Datum gesetzt, gilt der Raum unbegrenzt weiter.
	 */
	public gueltigBis: string | null = null;


	/**
	 * Leerer Standardkonstruktor.
	 */
	public constructor() {
		super();
	}

	/**
	 * Vergleicht, ob das aktuelle Objekt dasselbe ist wie ein anderes übergebenes Objekt.
	 *
	 * @param another das zu vergleichende Objekt
	 * @return true, falls die Objekte identisch sind, sonst false
	 */
	public equals(another: unknown | null): boolean {
		return (((another instanceof JavaObject) && (another.isTranspiledInstanceOf('de.svws_nrw.core.data.uv.UvRaum')))) && (this.id === (another as unknown as UvRaum).id);
	}

	/**
	 * Erzeugt den Hashcode zu Objekt auf Basis der ID.
	 *
	 * @return den HashCode
	 */
	public hashCode(): number {
		return JavaLong.hashCode((this.id));
	}

	/**
	 * Gibt eine String-Repräsentation des Objekts zurück.
	 *
	 * @return die String-Darstellung des Raums
	 */
	public toString(): string | null {
		return this.id + "-" + this.kuerzel + "-" + this.gueltigVon;
	}

	transpilerCanonicalName(): string {
		return 'de.svws_nrw.core.data.uv.UvRaum';
	}

	isTranspiledInstanceOf(name: string): boolean {
		return ['de.svws_nrw.core.data.uv.UvRaum'].includes(name);
	}

	public static readonly class = new Class<UvRaum>('de.svws_nrw.core.data.uv.UvRaum');

	public static transpilerFromJSON(json: string): UvRaum {
		const obj = JSON.parse(json) as Partial<UvRaum>;
		const result = new UvRaum();
		if (obj.id === undefined)
			throw new Error('invalid json format, missing attribute id');
		result.id = obj.id;
		if (obj.kuerzel === undefined)
			throw new Error('invalid json format, missing attribute kuerzel');
		result.kuerzel = obj.kuerzel;
		if (obj.gueltigVon === undefined)
			throw new Error('invalid json format, missing attribute gueltigVon');
		result.gueltigVon = obj.gueltigVon;
		result.gueltigBis = (obj.gueltigBis === undefined) ? null : obj.gueltigBis === null ? null : obj.gueltigBis;
		return result;
	}

	public static transpilerToJSON(obj: UvRaum): string {
		let result = '{';
		result += '"id" : ' + obj.id.toString() + ',';
		result += '"kuerzel" : ' + JSON.stringify(obj.kuerzel) + ',';
		result += '"gueltigVon" : ' + JSON.stringify(obj.gueltigVon) + ',';
		result += '"gueltigBis" : ' + ((obj.gueltigBis === null) ? 'null' : JSON.stringify(obj.gueltigBis)) + ',';
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

	public static transpilerToJSONPatch(obj: Partial<UvRaum>): string {
		let result = '{';
		if (obj.id !== undefined) {
			result += '"id" : ' + obj.id.toString() + ',';
		}
		if (obj.kuerzel !== undefined) {
			result += '"kuerzel" : ' + JSON.stringify(obj.kuerzel) + ',';
		}
		if (obj.gueltigVon !== undefined) {
			result += '"gueltigVon" : ' + JSON.stringify(obj.gueltigVon) + ',';
		}
		if (obj.gueltigBis !== undefined) {
			result += '"gueltigBis" : ' + ((obj.gueltigBis === null) ? 'null' : JSON.stringify(obj.gueltigBis)) + ',';
		}
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

}

export function cast_de_svws_nrw_core_data_uv_UvRaum(obj: unknown): UvRaum {
	return obj as UvRaum;
}

File: svws-core/src/main/java/de/svws_nrw/core/data/uv/UvRaum.java
package de.svws_nrw.core.data.uv;

import de.svws_nrw.transpiler.TranspilerDTO;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotNull;
import jakarta.xml.bind.annotation.XmlRootElement;

/**
 * Diese Klasse wird bei der Kommunikation über die Open-API-Schnittstelle verwendet.
 * Sie liefert die Informationen zu einem Raum innerhalb der Unterrichtsverteilung.
 */
@XmlRootElement
@Schema(description = "die Informationen zu einem Raum innerhalb der Unterrichtsverteilung.")
@TranspilerDTO
public class UvRaum {

	/** Die eindeutige ID des Raums. */
	@Schema(description = "die eindeutige ID des Raums", example = "4711")
	public long id = -1;

	/** Das Kürzel des Raums. */
	@Schema(description = "das Kürzel des Raums", example = "R204")
	public @NotNull String kuerzel = "";

	/** Die Beschreibung des Raumes. */
	@Schema(description = "die Beschreibung des Raumes", example = "Klassenraum der Klasse 07b")
	public String beschreibung = "";

	/** Die Größe des Raumes an Arbeitsplätzen für Schüler. */
	@Schema(description = "die Größe des Raumes an Arbeitsplätzen für Schüler", example = "30")
	public int groesse = -1;

	/** Die ID der Raumgruppe, falls der Raum zu einer solchen gehört. */
	@Schema(description = "die ID der Raumgruppe, falls der Raum zu einer solchen gehört", example = "102")
	public Long idRaumgruppe = null;

	/** Das Datum, ab dem der Raum gültig ist. */
	@Schema(description = "das Datum, ab dem der Raum gültig ist", example = "2025-08-01")
	public @NotNull String gueltigVon = "";

	/** Das Datum, bis wann der Raum gültig ist. Ist kein Datum gesetzt, gilt der Raum unbegrenzt weiter. */
	@Schema(description = "das Datum, bis wann der Raum gültig ist. Ist kein Datum gesetzt, gilt der Raum unbegrenzt weiter.", example = "2026-07-31")
	public String gueltigBis = null;

	/**
	 * Leerer Standardkonstruktor.
	 */
	public UvRaum() {
		// leer
	}

	/**
	 * Gibt eine String-Repräsentation des UvRaum-Objekts zurück.
	 *
	 * @return die String-Darstellung des Raums
	 */
	@Override
	public String toString() {
		return "UvRaum{id=" + id
				+ ", kuerzel=" + kuerzel
				+ ", gueltigVon=" + gueltigVon
				+ ", gueltigBis=" + gueltigBis
				+ "}";
	}

	/**
	 * Vergleicht, ob das aktuelle Objekt dasselbe ist wie ein anderes übergebenes Objekt.
	 *
	 * @param another das zu vergleichende Objekt
	 * @return true, falls die Objekte identisch sind, sonst false
	 */
	@Override
	public boolean equals(final Object another) {
		return (another instanceof final UvRaum a) && (this.id == a.id);
	}

	/**
	 * Erzeugt den Hashcode zu Objekt auf Basis der ID.
	 *
	 * @return den HashCode
	 */
	@Override
	public int hashCode() {
		return Long.hashCode(id);
	}

}
